File: src/main/java/com/intetics/lukyanenko/config/WebApp.java
package com.intetics.lukyanenko.config;

import org.springframework.web.WebApplicationInitializer;
import org.springframework.web.context.ContextLoaderListener;
import org.springframework.web.context.support.AnnotationConfigWebApplicationContext;
import org.springframework.web.filter.DelegatingFilterProxy;
import org.springframework.web.servlet.DispatcherServlet;

import javax.servlet.*;

public class WebApp
    implements WebApplicationInitializer {
   
    @Override
    public void onStartup(ServletContext servletContext) throws ServletException {
        // Root (entire application) context and listener for holding this context
        servletContext.addListener(new ContextLoaderListener(newSpringContext(RootConfig.class, RootSecurityConfig.class)));
        //
        // Main servlet and context
        Servlet servlet = new DispatcherServlet(newSpringContext(DispatcherServletConfig.class));
        ServletRegistration.Dynamic servletRegistration = servletContext.addServlet("dispatcher", servlet);
        servletRegistration.setLoadOnStartup(1);
        servletRegistration.addMapping("/");
        // Spring security filter
        Filter filter = new DelegatingFilterProxy();
        FilterRegistration.Dynamic filterRegistration = servletContext.addFilter("springSecurityFilterChain", filter);
        filterRegistration.addMappingForUrlPatterns(null, false, "/*");
    }
    
    private AnnotationConfigWebApplicationContext newSpringContext(Class<?>...configClassList) {
        AnnotationConfigWebApplicationContext rootContext = new AnnotationConfigWebApplicationContext();
        rootContext.register(configClassList);
        return rootContext;
    }
}

File: src/main/java/com/intetics/lukyanenko/config/WebApp.kt
package com.intetics.lukyanenko.config

import org.springframework.web.WebApplicationInitializer
import org.springframework.web.context.ContextLoaderListener
import org.springframework.web.context.support.AnnotationConfigWebApplicationContext
import org.springframework.web.filter.DelegatingFilterProxy
import org.springframework.web.servlet.DispatcherServlet

import javax.servlet.*

class WebApp : WebApplicationInitializer {

    override fun onStartup(servletContext: ServletContext) {
        // Root (entire application) context and listener for holding this context
        servletContext.addListener(ContextLoaderListener(newSpringContext(RootConfig::class.java, RootSecurityConfig::class.java)))
        //
        // Main servlet and context
        val servlet = DispatcherServlet(newSpringContext(DispatcherServletConfig::class.java))
        val servletRegistration = servletContext.addServlet("dispatcher", servlet)
        servletRegistration.setLoadOnStartup(1)
        servletRegistration.addMapping("/")
        // Spring security filter
        val filter = DelegatingFilterProxy()
        val filterRegistration = servletContext.addFilter("springSecurityFilterChain", filter)
        filterRegistration.addMappingForUrlPatterns(null, false, "/*")
    }

    private fun newSpringContext(vararg configClassList: Class<*>): AnnotationConfigWebApplicationContext {
        val rootContext = AnnotationConfigWebApplicationContext()
        rootContext.register(*configClassList)
        return rootContext
    }
}